File: comgongdaeng/DynamicProgramming/BOJ_2156.py
import sys

n = int(sys.stdin.readline())
wine = [0] + [int(sys.stdin.readline()) for _ in range(n)]
dp = [0] * (n + 1)
dp[1] = wine[1]
# 세 잔 연속 마실 수 없다?
# (1) 현재(i) 포도주를 먹지 않는다 == dp[i-1]
# (2) i-1, i 마시고 i-2는 마시지 않는다
# (3) i-2, i 마시고 i-1은 마시지 않는다
if n > 1:
    dp[2] = wine[1] + wine[2]

for i in range(3, n+1):
    dp[i] = max(dp[i-1], dp[i-3] + wine[i-1] + wine[i], dp[i-2] + wine[i])

print(dp[n])

File: keemsebin/DynamicProgramming/BOJ_2156.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");
const N = Number(input[0]);
let glass = [];
let Max;
for (let i = 1; i <= N; i++) {
  glass.push(Number(input[i]));
}
function wine(glass) {
  let dp = [];
  let wine = glass;
  dp[0] = wine[0]; //6
  dp[1] = wine[0] + wine[1]; // 6+10
  dp[2] = Math.max(wine[0] + wine[1], wine[0] + wine[2], wine[1] + wine[2]);
  for (let j = 3; j < N; j++) {
    dp[j] = Math.max(
      dp[j - 1], // 직전
      dp[j - 2] + wine[j], //
      dp[j - 3] + wine[j] + wine[j - 1]
    );
  }
  return dp[N - 1];
}
console.log(wine(glass));
